apps/widhid/wid.ts

```diff
@@ -0,0 +1,195 @@
+(() => {
+	const settings: Settings = require("Storage").readJSON("setting.json", true) || { HID: false } as Settings;
+	if (settings.HID !== "kbmedia") {
+		console.log("widhid: can't enable, HID setting isn't \"kbmedia\"");
+		return;
+	}
+
+	let anchor = {x:0,y:0};
+	let start = {x:0,y:0};
+	let dragging = false;
+	let activeTimeout: number | undefined;
+	let waitForRelease = true;
+
+	const onSwipe = ((_lr, ud) => {
+		if((Bangle as BangleExt).CLKINFO_FOCUS) return;
+
+		if(!activeTimeout && ud! > 0){
+			listen();
+			Bangle.buzz(20);
+		}
+	}) satisfies SwipeCallback;
+
+	const onDrag = (e => {
+		if((Bangle as BangleExt).CLKINFO_FOCUS) return;
+
+		if(e.b === 0){
+			// released
+			const wasDragging = dragging;
+			dragging = false;
+
+			if(waitForRelease){
+				waitForRelease = false;
+				return;
+			}
+
+			if(!wasDragging // i.e. tap
+			|| (Math.abs(e.x - anchor.x) < 2 && Math.abs(e.y - anchor.y) < 2))
+			{
+				toggle();
+				onEvent();
+				return;
+			}
+		}
+		if(waitForRelease) return;
+
+		if(e.b && !dragging){
+			dragging = true;
+			setStart(e);
+			Object.assign(anchor, start);
+			return;
+		}
+
+		const dx = e.x - start.x;
+		const dy = e.y - start.y;
+
+		if(Math.abs(dy) > 25 && Math.abs(dx) > 25){
+			// diagonal, ignore
+			setStart(e);
+			return;
+		}
+
+		// had a drag in a single axis
+		if(dx > 40){       next(); onEvent(); waitForRelease = true; }
+		else if(dx < -40){ prev(); onEvent(); waitForRelease = true; }
+		else if(dy > 30){  down(); onEvent(); setStart(e); }
+		else if(dy < -30){ up();   onEvent(); setStart(e); }
+	}) satisfies DragCallback;
+
+	const setStart = ({ x, y }: { x: number, y: number }) => {
+		start.x = x;
+		start.y = y;
+	};
+
+	const onEvent = () => {
+		Bangle.buzz(20); // feedback event sent
+		listen(); // had an event, keep listening for more
+	};
+
+	const listen = () => {
+		const wasActive = !!activeTimeout;
+		if(!wasActive){
+			suspendOthers();
+			waitForRelease = true; // wait for first touch up before accepting gestures
+			Bangle.on("drag", onDrag);
+			redraw();
+		}
+
+		if(activeTimeout) clearTimeout(activeTimeout);
+		activeTimeout = setTimeout(() => {
+			activeTimeout = undefined;
+
+			Bangle.removeListener("drag", onDrag);
+			resumeOthers();
+
+			redraw();
+		}, 3000);
+	};
+
+	const redraw = () => setTimeout(Bangle.drawWidgets, 50);
+
+	const connected = NRF.getSecurityStatus().connected;
```

Comment:
Out of interest, I've seen some scripts deleting variables - should I be deleting variables like `connected` to save memory, even if they're not captured by any closures? Or is there a different rationale behind that?

Reply:
In this case `connected` will hang around, so yes you could delete it or try and avoid using it (if not for memory usage, just to avoid having something called `connected` kicking around which doesn't reflect the current state!). 

If you've got a function defined which could possibly access a variable then Espruino will have to keep it - it's not smart enough to remove variables that aren't used in referencing functions.

Reply:
Gotya, makes sense - I'll do that what with this being a widget then, I reckon - thanks for the explanation